### PR TITLE
Support different topologies for containment queries

### PR DESCRIPTION
**Pull request by shai-doron**

### CODE REVIEW

**Comment by shai-doron** [comment]
@zivnevo, ready for your review.

File: network-config-analyzer/NetworkConfigQuery.py [hunk 1]
```diff
@@ -826,26 +824,35 @@ class ContainmentQuery(TwoNetworkConfigsQuery):
...
+            #  without it we fail on some calico tests
+            if only_captured and not captured1:
+                continue
+            _, conns2, _ = self.config2.allowed_connections(pair[0], pair[1])
+            if not conns1.contained_in(conns2):
+                output_result = f'{self.name1} is not contained in {self.name2}'
+                output_explanation = f'Allowed connections from {pair[0]} to {pair[1]} in {self.name1} ' \
+                                     f'are not a subset of those in {self.name2}\n'
+                output_explanation += conns1.print_diff(conns2, self.name1, self.name2)
+                return QueryAnswer(False, output_result, output_explanation)
```
Comment (zivnevo) [inline network-config-analyzer/NetworkConfigQuery.py:855]: We have to discuss this. I thought we can just check if `config1.peer_container` is contained in `config.peer_container` and then continue as before. Why isn't it the case?

**Review by zivnevo — CHANGES_REQUESTED** [review]

File: network-config-analyzer/NetworkConfigQuery.py [hunk 2]
```diff
@@ -868,25 +875,36 @@ def exec(self):
...
                                output_explanation=output_explanation, numerical_result=0)
         if contained_1_in_2.bool_result:
             return QueryAnswer(bool_result=False,
-                               output_result=f'NetworkPolicy set {self.name1} is a proper subset of {self.name2}.',
+                               output_result=f'network configuration set {self.name1} is a proper subset of {self.name2}.',
                                output_explanation=explanation_not_contained_other_self, numerical_result=2)
         # (contained_2_in_1)
         return QueryAnswer(bool_result=False,
-                           output_result=f'NetworkPolicy set {self.name2} is a proper subset of {self.name1}.',
+                           output_result=f'network configuration set {self.name2} is a proper subset of {self.name1}.',
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:893]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 3]
```diff
@@ -868,25 +875,36 @@ def exec(self):
...
             output_explanation = explanation_not_contained_self_other + '\n' + explanation_not_contained_other_self
             return QueryAnswer(bool_result=False,
-                               output_result=f'Neither set of NetworkPolicies {self.name1} and {self.name2}'
+                               output_result=f'Neither set of network configurations {self.name1} and {self.name2} '
                                              'is contained in the other.',
                                output_explanation=output_explanation, numerical_result=0)
         if contained_1_in_2.bool_result:
             return QueryAnswer(bool_result=False,
-                               output_result=f'NetworkPolicy set {self.name1} is a proper subset of {self.name2}.',
+                               output_result=f'network configuration set {self.name1} is a proper subset of {self.name2}.',
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:889]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 4]
```diff
@@ -868,25 +875,36 @@ def exec(self):
             contained_2_in_1.output_result + ':\n\t' + contained_2_in_1.output_explanation
         if contained_1_in_2.bool_result and contained_2_in_1.bool_result:
             return QueryAnswer(bool_result=False,
-                               output_result=f'The two sets of NetworkPolicies {self.name1} and {self.name2}'
+                               output_result=f'The two sets of network configurations {self.name1} and {self.name2} '
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:878]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 5]
```diff
@@ -868,25 +875,36 @@ def exec(self):
...
                                              'are semantically equivalent.',
                                numerical_result=3)
         if not contained_1_in_2.bool_result and not contained_2_in_1.bool_result:
             output_explanation = explanation_not_contained_self_other + '\n' + explanation_not_contained_other_self
             return QueryAnswer(bool_result=False,
-                               output_result=f'Neither set of NetworkPolicies {self.name1} and {self.name2}'
+                               output_result=f'Neither set of network configurations {self.name1} and {self.name2} '
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:884]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 6]
```diff
@@ -429,11 +429,11 @@ def __init__(self, config1, config2):
 
     def is_identical_topologies(self, check_same_policies=False):
         if self.config1.peer_container != self.config2.peer_container:
-            return QueryAnswer(False, 'The two NetworkPolicy sets are not defined over the same set of endpoints, '
+            return QueryAnswer(False, 'The two configurations have different network topologies, '
                                       'and are thus not comparable.')
         if check_same_policies and self.config1.policies == self.config2.policies and \
                 self.config1.profiles == self.config2.profiles:
-            return QueryAnswer(True, f'{self.name1} and {self.name2} have exactly the same set of policies')
+            return QueryAnswer(True, f'{self.name1} and {self.name2} have the same network topology and set of policies')
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:436]: done

File: network-config-analyzer/NetworkConfigQuery.py [hunk 7]
```diff
@@ -868,25 +875,36 @@ def exec(self):
...
-                           output_result=f'NetworkPolicy set {self.name2} is a proper subset of {self.name1}.',
+                           output_result=f'network configuration set {self.name2} is a proper subset of {self.name1}.',
                            output_explanation=explanation_not_contained_self_other, numerical_result=1)
 
 
+class PermitsQuery(TwoNetworkConfigsQuery):
+    """
+    Checking whether the connections explicitly allowed by config1 are explicitly allowed by config2
+    """
+    def exec(self, only_captured=False):
```
[the remark this answers is not in the document]
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:901]: done

File: network-config-analyzer/NetworkConfigQuery.py (hunk 1 above)
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py:855]: done per discussion

File: network-config-analyzer/NetworkConfigQuery.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zivnevo) [reply network-config-analyzer/NetworkConfigQuery.py:893]: where is this resolved? did you push your changes?

**Comment by shai-doron** [comment]
@zivnevo, further items to discuss once you finish reviewing:
 1. In case of config with no policy, i.e. `networkPolicyList: []`, there's a problem in `_allowed_xgress_conns()` that doesn't take into account no k8s policy. Problem is that `policy-type` is not defined by default as `k8s`. This is the reason why some of the new tests fail (left it intentionally so it must be fixed)
 2. Do we want to enhance Containment (and maybe other queries like permit/forbid) to find and return all evidences?

File: network-config-analyzer/NetworkConfigQuery.py [hunk 8]
```diff
@@ -429,11 +429,12 @@ def __init__(self, config1, config2):
...
     def is_identical_topologies(self, check_same_policies=False):
         if self.config1.peer_container != self.config2.peer_container:
-            return QueryAnswer(False, 'The two NetworkPolicy sets are not defined over the same set of endpoints, '
-                                      'and are thus not comparable.')
+            return QueryAnswer(False, 'The two configurations have different network topologies '
+                                      'and thus are not comparable')
         if check_same_policies and self.config1.policies == self.config2.policies and \
                 self.config1.profiles == self.config2.profiles:
-            return QueryAnswer(True, f'{self.name1} and {self.name2} have exactly the same set of policies')
+            return QueryAnswer(True, f'{self.name1} and {self.name2} have the same network'
```
Comment (zivnevo) [inline network-config-analyzer/NetworkConfigQuery.py:436]:
```suggestion
            return QueryAnswer(True, f'{self.name1} and {self.name2} have the same network '
```
A missing space

File: network-config-analyzer/ExecutionAssist.py [hunk 9]
```diff
@@ -131,7 +131,7 @@ def execute(self):
         print()
         full_result = ContainmentQuery(self.permit_config, self.base_config).exec(True)
         if not full_result.bool_result:
-            print(f'{self.base_config.name} does not permit connections specified in {self.permit_config.name}:')
+            print(f'{self.base_config.name} does not explicitly permit connections specified in {self.permit_config.name}:')
```
Comment (zivnevo) [inline network-config-analyzer/ExecutionAssist.py:134]:
```suggestion
            print(f'{self.base_config.name} does not permit connections specified in {self.permit_config.name}:')
```
It's also ok if `self.base_config` permits the connections implicitly.

**Review by zivnevo — CHANGES_REQUESTED** [review]
Just these two minor changes, and we are ready to merge

**Comment by zivnevo** [comment]
> 
> 
> @zivnevo, further items to discuss once you finish reviewing:
> 
>     1. In case of config with no policy, i.e. `networkPolicyList: []`, there's a problem in `_allowed_xgress_conns()` that doesn't take into account no k8s policy. Problem is that `policy-type` is not defined by default as `k8s`. This is the reason why some of the new tests fail (left it intentionally so it must be fixed)
> 
>     2. Do we want to enhance Containment (and maybe other queries like permit/forbid) to find and return all evidences?

1. This is a very good point. We set a NetworkConfig's `type` according to the policies it contains. If it contains no policies we don't really know what type of config this is. How about adding a new optional field `type` to the definition of a NetworkConfig? This field should be useful only if there are no policies. Otherwise,  we'll only check they type specified matches the policies (maybe this is also useful as a sanity check).  
Also, please let me know if you can merge (without using your administrator power) while some tests are failing.
2. Yes, but I think we want this to be optional. Adi is preparing an output-configuration class. Let's wait for it to be ready, and take the option from there.

File: network-config-analyzer/NetworkConfigQuery.py (hunk 8 above)
Comment (shai-doron) [reply network-config-analyzer/NetworkConfigQuery.py]: done

File: network-config-analyzer/ExecutionAssist.py (hunk 9 above)
Comment (shai-doron) [reply network-config-analyzer/ExecutionAssist.py]: done

**Review by zivnevo — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**